Sources/LeafKit/LeafSyntax/LeafSyntax.swift

```diff
@@ -68,28 +172,110 @@ extension Syntax {
         }
     }
 
-    public struct Extend {
+    public struct Extend: BodiedSyntax {
         public let key: String
         public private(set) var exports: [String: Export]
+        private var externalsSet: Set<String>
+        private var importSet: Set<String>
 
         public init(_ params: [ParameterDeclaration], body: [Syntax]) throws {
             guard params.count == 1 else { throw "extend only supports single param \(params)" }
             guard case .parameter(let p) = params[0] else { throw "extend expected parameter type, got \(params[0])" }
             guard case .stringLiteral(let s) = p else { throw "import only supports string literals" }
             self.key = s
+            self.externalsSet = .init(arrayLiteral: self.key)
+            self.importSet = .init()
+            self.exports = [:]
 
-            var exports: [String: Export] = [:]
             try body.forEach { syntax in
                 switch syntax {
                     // extend can ONLY export, raw space in body ignored
                     case .raw: return
                     case .export(let export):
-                        exports[export.key] = export
+                        guard !export.externals().contains(self.key) else {
+                            throw LeafError(.cyclicalReference(self.key, [self.key]))
+                        }
+                        self.exports[export.key] = export
+                        externalsSet.formUnion(export.externals())
+                        importSet.formUnion(export.imports())
                     default:
                         throw "unexpected token in extend body: \(syntax).. use raw space and `export` only"
                 }
             }
+        }
+        
+        internal init(key: String, exports: [String : Syntax.Export], externalsSet: Set<String>, importSet: Set<String>) {
+            self.key = key
             self.exports = exports
+            self.externalsSet = externalsSet
+            self.importSet = importSet
+        }
+        
+        func externals() -> Set<String> {
+            return externalsSet
+        }
+        func imports() -> Set<String> {
+            return importSet
+        }
+        
+        func inlineRefs(_ externals: [String: LeafAST], _ imports: [String : Syntax.Export]) -> [Syntax] {
+            var newExports = [String: Export]()
+            var newImports = imports
+            var newExternalsSet = Set<String>()
+            var newImportSet = Set<String>()
```

Comment:
Syntax.Extend.inlineRefs is the main behavioral interface for extending; at the top level of a document, no imports exist yet. Once an Extend is hit, the Extend will then inject any external references into it's own Exports, rebuilding those first, and then loading the external template it references if available, directly rebuilding the external Syntax as it assembles the current resolvable state of the Syntax - then passing the complete [Syntax] back to the top level of the AST (or to the higher level of the AST tree if nested within another Syntax).